### ISSUE TITLE
Tips for reliable "get" stream needed

### ISSUE THREAD

**Issue by zhlong73** [facebook/mcrouter#90]
Hi team,

I have mcrouter running (mcrouter -b -p 5000 -f /var/mcrouter.cfg) with a very simple config as below:
{
    "pools": {
        "replication_pool": {
         "servers": [ "192.168.10.51:11212", "192.168.10.52:11212", "192.168.10.60:11212"]
        }
    },
    "route": {
        "type": "OperationSelectorRoute",
        //"default_policy" : "PoolRoute|replication_pool",
        "operation_policies": {
            "add": "AllSyncRoute|Pool|replication_pool",
            "delete": "AllSyncRoute|Pool|replication_pool",
            "set": "AllFastestRoute|Pool|replication_pool",
            "get": "MissFailoverRoute|Pool|replication_pool"
        }
    }
}

Application written in PHP and executed by hhvm (version 3.9.1) interacts with mcrouter through memcached class:
$mcHost = "127.0.0.1";
$mcPort = 5000;
$mcrouter = new Memcached();
$mcrouter->addServer($mcHost, $mcPort);

I shut down 1 memcached instance, and did a bunch of "set" and "get" requests. "Set" ones worked perfect. All the kvp get replicated to the healthy memcached instances, but "get" requests always returned false with "Not Found" as the result message. I did a bunch of tcpdump to debug get requests:

1) capture from source server on port 5000 from loopback interface (tcpdump -i lo -n dst port 5000):
14:17:33.344918 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [P.], seq 4797892:4806088, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142808], length 8196
14:17:33.344940 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4806088:4822472, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142808], length 16384
14:17:33.344957 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4822472:4838856, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142808], length 16384
14:17:33.344989 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4838856:4855240, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142808], length 16384
14:17:33.344998 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [P.], seq 4855240:4855264, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 24
14:17:33.345006 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4855264:4871648, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 16384
14:17:33.345017 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4871648:4888032, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 16384
14:17:33.345027 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4888032:4904416, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 16384
14:17:33.345033 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [P.], seq 4904416:4904440, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 24
14:17:33.345040 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [.], seq 4904440:4920824, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 16384
14:17:33.345054 IP 127.0.0.1.44562 > 127.0.0.1.5000: Flags [P.], seq 4920824:4929028, ack 2684867, win 4113, options [nop,nop,TS val 430142816 ecr 430142816], length 820 

I can see application send "get" requests to mcrouter instance.

2) capture from source server on eth1 with dst port 11212 (tcpdump -i eth1 -n dst port 11212) just returned nothing. It seems that mcrouter didn't fan-out get requests to memcached instances at all. Same capture returns the full trace for set requests.

3) capture from servers with healthy memcached running returned nothing.

4) get using telnet just worked (telnet localhost 5000 and do a few gets).

Could you please let me know if there is any thing obviously wrong with my setup and advise some further steps I can do to troubleshoot the issue? 

Thanks,
~L

**Comment by zhlong73** [comment]
BTW,
1) All the hosts are running ubuntu 12.04 with is sad...
2) mcrouter/folly and everything are just built with most up-to-date code from github
3) function const char *AsyncSSLSocket::getSSLCertSigAlgName() const in https://github.com/facebook/folly/blob/master/folly/io/async/AsyncSSLSocket.cpp#L810 is not declared in AsyncSSLSocket.h, which caused build error on gcc 4.8.

**Comment by zhlong73** [comment]
Alright, issue get fixed after taking a further look at mcrouter's code. The commands sent from application were "gets" instead of "get"... Adding following line in mcrouter.cfg fixed the problem.
"gets": "MissFailoverRoute|Pool|replication_pool" 

It will be great if further explanation could be given.

**Comment by pavlo-fb** [comment]
Hi @zhlong73!
First of all, `get` and `gets` are different operations. `gets` returns `cas` token; `get` is a simple retrieval command. Mcrouter doesn't convert `get` to `gets`, it looks like the application itself sent `gets`. The right way to go will be to figure out why application uses `set` and then retrieves a value with `gets`, it doesn't look correct, since `gets` should be used together with `cas`. More information about get/gets is here: https://github.com/memcached/memcached/blob/master/doc/protocol.txt

Second, mcrouter doesn't convert `get` to `gets`. It sends the same operation it receives.

And regarding broken build - sorry about that, folly was fixed (see https://github.com/facebook/folly/commit/b4c40ee9da8183d78bf5506562e9c241f659f71b).

**Comment by zhlong73** [comment]
Hi Pavlo,

Thanks for the reply! From http://docs.hhvm.com/manual/en/book.memcached.php, "gets" is not a method that being exposed to the application. I will take a further look at memcached extention implemented in HHVM. Thank you!

~L

**Comment by pavlo-fb** [comment]
Apparently you pass `cas_token` to `Memcached::get`, check http://docs.hhvm.com/manual/en/memcached.get.php

**Comment by zhlong73** [comment]
Thanks Pavlo. I agree the issue is most likely to be on the application side. I didn't explicitly set the cas_token for either get or set. As I posted above, the code is something like:
    $mcHost = "127.0.0.1";
    $mcPort = 5000;
    $mcrouter = new Memcached();
    $mcrouter->addServer($mcHost, $mcPort);

Afterwards, I did  $mcrouter->set() for 10k keys with values around 2mb for each key. And did $mcrouter->get() for these keys in a for loop. There is no other parameters provided along with the keys for get and set cmd. The php execution engine is hhvm-3.9.1.  

The interesting part is I only see this conversion (get->gets) when the network condition is good (application server running on the same rack with servers hosting memcached). If I move the pool to instances hosted in AWS (higher RTT), the requests application sent was "get".
